File: Scripts/MainScript.R
#| Animal Behaviour Classification using Machine Learning
#| 
#| This script implements a comparative analysis of machine learning approaches
#| for classifying animal behaviors from accelerometer data. It compares:
#| - One-Class Classification (OCC)
#| - Binary Classification
#| - Multi-class Classification


# dataset_name <- "Vehkaoja_Dog"
dataset_name <- "Ladds_Seal"
base_path <- "C:/Users/oaw001/OneDrive - University of the Sunshine Coast/AnomalyDetection"
# base_path <- "C:/Users/PC/OneDrive - University of the Sunshine Coast/AnomalyDetection"

# Set Up ------------------------------------------------------------------
# User defined variables for each dataset
#' @param sample_rate Sampling frequency in Hz
#' @param window_length Window size in seconds
#' @param overlap_percent Overlap between windows as a whole number
#' @param target_activities Behaviours to classify in the OCC and Binary models
window_settings <- list(
  Vehkaoja_Dog = list(
    sample_rate = 100,
    window_length = 1,
    overlap_percent = 50,
    target_activities = c("Walking", "Eating", "Shaking", "Lying chest")
  ),
  Ladds_Seal = list(
    sample_rate = 25,
    window_length = 1,
    overlap_percent = 50,
    target_activities = c("swimming", "still", "chewing", "facerub")
  )
)

sample_rate <- window_settings[[dataset_name]]$sample_rate
window_length <- window_settings[[dataset_name]]$window_length
overlap_percent <- window_settings[[dataset_name]]$overlap_percent
target_activities <- window_settings[[dataset_name]]$target_activities

# Load Required Packages --------------------------------------------------
library(pacman)
p_load(
  caret,           # Classification and regression training
  data.table,      # Fast data manipulation
  e1071,           # SVM implementation
  future,          # Parallel processing
  future.apply,    # Parallel apply functions
  parallelly,      # Parallel processing utilities
  plotly,          # Interactive plotting
  purrr,           # Functional programming tools
  pROC,            # ROC curve analysis
  rBayesianOptimization,  # Bayesian optimization
  MLmetrics,       # Machine learning metrics
  ranger,          # Fast random forests
  tsfeatures,      # Time series feature extraction
  tidyverse,       # Data manipulation and visualization
  zoo,             # Time series functions
  patchwork        # Combine plots
)

# Global Variables -------------------------------------------------------
all_axes <- c("Accelerometer.X", "Accelerometer.Y", "Accelerometer.Z")
label_columns <- c("Activity", "Time", "ID")
test_proportion <- 0.2
validation_proportion <- 0.2
features_type <- c("timeseries", "statistical")
balance <- "stratified_balance"

# Custom Functions -------------------------------------------------------
function_files <- c(
  "FeatureGenerationFunctions.R",      # Feature extraction
  "FeatureSelectionFunctions.R",       # Feature selection methods
  "ModelTuningFunctions.R",            # Model optimisation
  "PerformanceCalculationFunctions.R", # Functions for perofmrance and baselines
  "OtherFunctions.R",                  # Utility functions
  "PlotFunctions.R"                    # Generating performance plots
)

invisible(lapply(function_files, function(file) {
  source(file.path(base_path, "Scripts", "Functions", file))
}))

# Analysis Pipeline ----------------------------------------------------

# 1. Split Data into Training and Test Sets
source(file.path(base_path, "Scripts", "SplitTestData.R"))

# 2. Data Exploration and Clustering
source(file.path(base_path, "Scripts", "DataExploration.R"))

# 3. Data Preprocessing, Feature Engineering, and re-clustering
source(file.path(base_path, "Scripts", "Preprocessing.R"))
source(file.path(base_path, "Scripts", "ClusteringBehaviours.R"))

# 4. Hyperparameter Optimization
source(file.path(base_path, "Scripts", "HpoOptimisation.R"))

# 5. Model Evaluation
source(file.path(base_path, "Scripts", "TestBestModels.R"))

# 6. Results Visualization and Comparison
source(file.path(base_path, "Scripts", "PlottingPerformance.R"))
source(file.path(base_path, "Scripts", "PlotPredictions.R"))

File: Scripts/Python/MainScript.py
"""
Open Set Recognition for Animal Behaviour Classification from Accelerometer Data

This program implements a comparative analysis of machine learning approaches
for classifying animal behaviours from accelerometer data.
It compares SVM and Random Forest based systems for:
- One-Class Classification (OCC)
- Binary Classification
- Multi-class Classification
"""

# define the variables
# BASE_PATH = "C:/Users/oaw001/OneDrive - University of the Sunshine Coast/AnomalyDetection"
BASE_PATH = "C:/Users/PC/OneDrive - University of the Sunshine Coast/AnomalyDetection"

target_activities = {
    "Vehkaoja_Dog": ["Walking", "Eating", "Shaking", "Lying chest"],
    "Ladds_Seal": ["swimming", "still", "chewing"],
    "Ferdinandy_Dog": ["walk", "eat", "lay"]  # only 3 target activities
    }

def main():
    # Import and run the create_datasets module
    import sys
    import os
    from pathlib import Path
    from itertools import product

    # import HpoOptimisation
    # import CreateDatasets
    import TrainModel
    import TestModelOpen
    # import PlotResults
    import CombineResults
    
    # Add the project root directory to Python path
    sys.path.append(BASE_PATH)

    DATASET_NAME = "Ferdinandy_Dog" 
    TARGET_ACTIVITIES = target_activities[DATASET_NAME]

    # parameters for the experiments
    FOLDS = [1, 2, 3, 4, 5]
    TRAINING_SETS = ['some', 'target', 'all']
    MODEL_TYPES = ['multi', 'binary', 'oneclass']
    BEHAVIOUR_SETS = ['Other', 'Activity']
    THRESHOLD_OPTIONS = [False, True]

    # Remove all files in Metrics folders for each fold
    for fold in FOLDS:
        metrics_path = Path(f"{BASE_PATH}/Output/fold_{fold}/Testing/Metrics/")
        if metrics_path.exists():
            print(f"Removing files in {metrics_path}")
            for file in metrics_path.glob('*.csv'):
                file.unlink()

    for FOLD in FOLDS:
        for TRAINING_SET in TRAINING_SETS:
            for MODEL_TYPE in MODEL_TYPES:
                
                if MODEL_TYPE == 'multi':
                    for BEHAVIOUR_SET in BEHAVIOUR_SETS:
                        if BEHAVIOUR_SET.lower() == 'activity':
                            for THRESHOLDING in THRESHOLD_OPTIONS:
                                print(f"Thresholding: {THRESHOLDING}")
                                # train the models the same
                                TrainModel.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                                TARGET_ACTIVITIES, BEHAVIOUR_SET, THRESHOLDING, 
                                                FOLD=FOLD)
                                if THRESHOLDING:
                                    TestModelOpen.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                                TARGET_ACTIVITIES, BEHAVIOUR_SET, THRESHOLDING, 
                                                REASSIGN_LABELS=True, FOLD=FOLD)
                                else:
                                    TestModelOpen.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                                TARGET_ACTIVITIES, BEHAVIOUR_SET, THRESHOLDING, 
                                                REASSIGN_LABELS=True, FOLD=FOLD)
                                
                        else: # BEHAVIOUR_SET == 'Other'
                            TrainModel.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                                TARGET_ACTIVITIES, BEHAVIOUR_SET, THRESHOLDING = False, 
                                                FOLD=FOLD)

                            TestModelOpen.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                            TARGET_ACTIVITIES, BEHAVIOUR_SET, THRESHOLDING = False, 
                                            REASSIGN_LABELS=True, FOLD=FOLD)

                else:
                    # divide and conquer methods
                    TrainModel.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                                TARGET_ACTIVITIES, BEHAVIOUR_SET = None, THRESHOLDING = False, 
                                                FOLD=FOLD)
                    TestModelOpen.main(BASE_PATH, DATASET_NAME, TRAINING_SET, MODEL_TYPE, 
                                            TARGET_ACTIVITIES, BEHAVIOUR_SET = None, THRESHOLDING = False, 
                                            REASSIGN_LABELS=True, FOLD=FOLD)

    # First run the comparisons to generate data
    CombineResults.main(BASE_PATH)

    # PlotResults.main(BASE_PATH, "Ferdinandy_Dog")

    # plot the volume to performance graph
    # import PlotVolumePerformance
    # PlotVolumePerformance.main(BASE_PATH)

    # import PlotFullMulticlass
    # PlotFullMulticlass.generate_full_class_data(BASE_PATH, target_activities, FOLD = 2)
    # PlotFullMulticlass.main(BASE_PATH, target_activities, FOLD = 2)

if __name__ == "__main__":
    main()
